Update coherence and orphan rules documentation to match RFC 2451

ohadravid opened the pull request:
As part of stabilizing [RFC 2451](https://rust-lang.github.io/rfcs/2451-re-rebalancing-coherence.html) (re-rebalance coherence) in https://github.com/rust-lang/rust/pull/65879, I wanted to update the reference to match it.

I copied the relevant parts of the RFC, with a slight rewording by @nikomatsakis ([comment](https://github.com/rust-lang/rust/issues/63599#issuecomment-522032863)).

Centril commented on src/items/coherence.md:
```diff
@@ -0,0 +1,79 @@
+## What is coherence and why do we care?
+
+Coherence means that for any given trait and type, there is one specific 
```
Can you try to remove the motivational stuff from this text? The reference's only aim is to give a definition / specification of behavior, not explain why things matter and why we care.

Centril commented on src/items/coherence.md:
```diff
@@ -0,0 +1,79 @@
+In languages without coherence, the compiler has to have some way to choose
+which implementation to use when multiple implementations could apply. Scala
+does this by having complex scope resolution rules for "implicit" parameters.
+Haskell (when a discouraged flag is enabled) does this by picking an impl
```
This entire paragraph is irrelevant. Haskell (mainly) and other languages are only interesting for the purposes of "influences".

Centril commented on src/items/coherence.md:
```diff
@@ -0,0 +1,79 @@
+    `Ti`)
+
+We only restrict the appearance of *uncovered* type parameters. Once again, it is
+important to note that for the purposes of coherence, `#[fundamental]` types are
```
`#[funtamental]` is not part of the stable language definition. From the POV of the reference, some types and traits are just fundamental. We can specify with a note how rustc achieves this. What is part of the language definition is the actual set of fundamental types and traits. These should be itemized here.

Centril commented on src/items/coherence.md:
```diff
@@ -0,0 +1,79 @@
+Vec<T>` is not a blanket impl, as all instances of `T` which appear in this impl
+are covered by `Vec`.
+
+Fundamental Type: A type for which you cannot add a blanket impl backwards
```
The social aspect of fundamental types is not important in the reference so the first sentence can be removed and the other ones restructured a bit.

Centril commented on src/items/coherence.md:
```diff
@@ -0,0 +1,79 @@
+compatibly. This includes `&`, `&mut`, and `Box`. Any time a type `T` is
+considered local, `&T`, `&mut T`, and `Box<T>` are also considered local.
+Fundamental types cannot cover other types. Any time the term "covered type" is
+used, the `T` in `&T`, `&mut T`, and `Box<T>` is not considered covered.
```
Seems like some of these terms should be added to the glossary?

Centril commented on src/items/coherence.md:
```diff
@@ -0,0 +1,79 @@
+Haskell (when a discouraged flag is enabled) does this by picking an impl
+arbitrarily.
+
+Rust's solution is to enforce that there is only one impl to choose from at all.
```
Most of this paragraph is not important either.

Centril commented on src/items/coherence.md:
```diff
@@ -0,0 +1,79 @@
+## What is coherence and why do we care?
```
I think this text should mainly explain that overlap is prevented (and the rules for that) and that orphans are not allowed (and the rules for that).

Question to @nikomatsakis: is the orphan rule actually part of preventing overlap in the compiler?

ohadravid commented:
Hi @Centril!

I made the changes you requested, and since all the terms seemed relevant I moved all of them to the glossary, making the coherence explanation much shorter so I moved it back to where it was. If you think this is too much for the glossary I can also move them back/move only some of them/etc.

You question to Niko about the orphan rule/overlap is still relevant, because the old text is not too specific there either (I changed it back to the way it was).

One question about defining fundamental types: I looked for the attribute, and saw that `Pin` was added since the RFC (So I added it to the list).
`Generator` is also there, but I didn't add it because it unstable.
So should there be a sentence like "the lang may define additional types which are considered fundamental"?

nikomatsakis commented:
@ohadravid thanks for doing this!

Centril commented on src/glossary.md:
```diff
@@ -30,6 +30,14 @@ items are defined in [implementations] and declared in [traits]. Only
+Any implementation where a type appears [uncovered](#uncovered-type). `impl<T> Foo
+for T`, `impl<T> Bar<T> for T`, `impl<T> Bar<Vec<T>> for T`, and `impl<T> Bar<T>
+for Vec<T>` are considered blanket impls. However, `impl<T> Bar<Vec<T>> for
+Vec<T>` is not a blanket impl, as all instances of `T` which appear in this impl
```
```suggestion
Vec<T>` is not a blanket impl, as all instances of `T` which appear in this `impl`
```

Centril commented on src/glossary.md:
```diff
@@ -65,6 +73,13 @@ For example, `2 + (3 * 4)` is an expression that returns the value 14.
+Includes `&`, `&mut`, `Box` and `Pin`. Any time a type `T` is
+considered [local](#local-type), `&T`, `&mut T`, `Box<T>` and `Pin<T>` are also considered local.
+Fundamental types cannot [cover](#uncovered-type) other types. Any time the term "covered type" is
+used, the `T` in `&T`, `&mut T`, `Box<T>` and `Pin<T>` is not considered covered.
```
```suggestion
used, the `T` in `&T`, `&mut T`, `Box<T>`, and `Pin<T>` is not considered covered.
```

Centril commented on src/glossary.md:
```diff
@@ -65,6 +73,13 @@ For example, `2 + (3 * 4)` is an expression that returns the value 14.
 
+### Fundamental Type
+
+Includes `&`, `&mut`, `Box` and `Pin`. Any time a type `T` is
```
Some points & questions (which @nikomatsakis may help with):

- `&` and friends are not types; they are type constructors.

- Suppose we have `#[fundamental] struct W<P0, P1> { ... }`. How would `W<A0, A1>` behave with respect to locality?  In this case the addition of a second type parameter is intentional so that we can suss out the *essence* of how fundamental types behave. It seems lifetimes have no effect on locality.

- Before itemizing the list of fundamental types it would be good to give a definition of how such types behave.

- What about fundamental `trait`s?

Centril commented on src/glossary.md:
```diff
@@ -87,6 +102,19 @@ A variable is initialized if it has been assigned a value and hasn't since been
 moved from. All other memory locations are assumed to be uninitialized. Only
 unsafe Rust can create such a memory without initializing it.
 
+### Local Trait
```
```suggestion
### Local trait
```

Centril commented on src/glossary.md:
```diff
@@ -87,6 +102,19 @@ A variable is initialized if it has been assigned a value and hasn't since been
 
+### Local Trait
+
+A trait which was defined in the current crate. Whether a trait is
```
```suggestion
A `trait` which was defined in the current crate. A trait definition is local
or not independent of applied type arguments. That is, given `trait Foo<T, U>`,
```

Centril commented on src/glossary.md:
```diff
@@ -87,6 +102,19 @@ A variable is initialized if it has been assigned a value and hasn't since been
+### Local Trait
+
+A trait which was defined in the current crate. Whether a trait is
+local or not has nothing to do with type parameters. Given `trait Foo<T, U>`,
```
```suggestion
```

Centril commented on src/glossary.md:
```diff
@@ -87,6 +102,19 @@ A variable is initialized if it has been assigned a value and hasn't since been
+
+A trait which was defined in the current crate. Whether a trait is
+local or not has nothing to do with type parameters. Given `trait Foo<T, U>`,
+`Foo` is always local, regardless of the types used for `T` or `U`.
```
```suggestion
`Foo` is always local, regardless of the types substituted for `T` and `U`.
```

Centril commented on src/glossary.md:
```diff
@@ -87,6 +102,19 @@ A variable is initialized if it has been assigned a value and hasn't since been
+
+### Local Type
+
+A struct, enum, or union which was defined in the current crate.
```
```suggestion
A `struct`, `enum`, or `union` which was defined in the current crate.
```

Centril commented on src/glossary.md:
```diff
@@ -87,6 +102,19 @@ A variable is initialized if it has been assigned a value and hasn't since been
+### Local Type
+
+A struct, enum, or union which was defined in the current crate.
+This is not affected by type parameters. `struct Foo` is considered local, but
```
```suggestion
This is not affected by applied type arguments. `struct Foo` is considered local, but
```

Centril commented on src/glossary.md:
```diff
@@ -87,6 +102,19 @@ A variable is initialized if it has been assigned a value and hasn't since been
+A struct, enum, or union which was defined in the current crate.
+This is not affected by type parameters. `struct Foo` is considered local, but
+`Vec<Foo>` is not. `LocalType<ForeignType>` is local. Type aliases and trait
+aliases do not affect locality.
```
Trait aliases do not exist on stable.

Centril commented on src/glossary.md:
```diff
@@ -158,6 +186,12 @@ It allows a type to make certain promises about its behavior.
 
+### Uncovered Type
+
+A type which does not appear as a parameter to another type. For example,
```
```suggestion
A type which does not appear as an argument to another type. For example,
```

Centril commented on src/glossary.md:
```diff
@@ -158,6 +186,12 @@ It allows a type to make certain promises about its behavior.
+
+A type which does not appear as a parameter to another type. For example,
+`T` is uncovered, but the `T` in `Vec<T>` is covered. This is only relevant for
+type parameters.
```
```suggestion
type arguments.
```

Centril commented on src/items/implementations.md:
```diff
@@ -152,29 +152,31 @@ impl Shape for Circle {
+#### Orphan rules
 
-1.  The trait being implemented is defined in the same crate.
+Given `impl<P1..=Pn> Trait<T1..=Tn> for T0`, an impl is valid only if at
```
```suggestion
Given `impl<P1..=Pn> Trait<T1..=Tn> for T0`, an `impl` is valid only if at
```

Centril commented on src/items/implementations.md:
```diff
@@ -152,29 +152,31 @@ impl Shape for Circle {
+  - No [uncovered type] parameters `P1..=Pn` may appear in `T0..Ti` (excluding
+    `Ti`)
+
+We only restrict the appearance of *uncovered* type parameters. Once again, it is
```
```suggestion
Only the appearance of *uncovered* type parameters is restricted.
```

Centril commented on src/items/implementations.md:
```diff
@@ -152,29 +152,31 @@ impl Shape for Circle {
+    `Ti`)
+
+We only restrict the appearance of *uncovered* type parameters. Once again, it is
+important to note that for the purposes of coherence, [fundamental types] are
```
```suggestion
Note that for the purposes of coherence, [fundamental types] are
```

Centril commented:
> If you think this is too much for the glossary I can also move them back/move only some of them/etc.

No it's good. :)

>  So should there be a sentence like "the lang may define additional types which are considered fundamental"?

I think we sorta want to give a description of behavior for any arbitrary fundamental type that is semi-syntactic in that it talks about type arguments and type constructors and how those relate wrt. locality and such. Then we can itemize a list of fundamental type constructors.

ohadravid commented:
I applied all your edits @Centril , also edited some more title case -> sentence case and such.

I changed the fundamental types part to use the type constructor wording, but I leave it to Niko to define what they are :)

I did use the order of: what are fundamentals, itemized list, locality.
Explaining it with locality first didn't sound right (you have to use at least `&T` and `Box<T>` when explaining locality, because of the different syntax, and than it seems like `&mut T` and `Pin` are somehow different..) but that's all more like a placeholder until we have the actual description.

ohadravid commented:
@Centril @nikomatsakis I used Niko's definitions from [RFC 1023](https://github.com/rust-lang/rfcs/blob/master/text/1023-rebalancing-coherence.md#errors-from-cargo-and-the-fundamental-attribute).
They seem to fit the reference, but I wasn't too sure if more explanation is needed. I also added the definition of fundamental traits for completeness.

Centril commented:
Filed https://github.com/rust-lang/reference/issues/714 for any follow-ups to the more formal spec.

Centril commented:
Thanks!
